docs/source/tutorial_torch_generator_agent.md

```diff
@@ -154,3 +153,79 @@ parlai train_model -m examples/seq2seq \
+
+Let us walk though an example:
+
+```python
+    def compute_loss(self, batch, return_output=False):
+        # first compute our regular forced decoding loss
+        token_loss, model_output = super().compute_loss(batch, return_output=True)
+        # compute our additional losses, you can replace this with your losses.
+        notnull = batch.label_vec.ne(self.NULL_IDX)
+        num_tokens = notnull.long().sum().item()
+
+        encoder_states = model_output[2]
+        ctx_know_attn = encoder_states[2]
+        # knowledge_alpha is the loss weight between the decoding loss and custom losses
+        if self.knowledge_alpha == 0.0:
+            loss = token_loss
+        else:
+            # calculate additional losses, you can send your extra field
+            # (like reward in RL tasks) in the batch
+            # we will cover this later
+            _, know_pred = ctx_know_attn.max(1)
+            know_acc = (know_pred == batch.cs_ids).float().sum().item()
+            know_chance = batch.ck_mask.sum(1).float().reciprocal().sum().item()
+            # don't forget to log your losses into metrics
+            self.metrics['know_chance'] += know_chance
+            self.metrics['bsz'] += batch.text_vec.size(0)
+            self.metrics['know_acc'] += know_acc
+            know_loss = th.nn.functional.cross_entropy(
+                ctx_know_attn, batch.cs_ids, reduction='mean'
+            )
+            self.metrics['know_loss'] += know_loss.item() * batch.text_vec.size(0)
+            know_loss /= num_tokens
+            # combine them
+            loss = (
+                1 - self.knowledge_alpha
+            ) * token_loss + self.knowledge_alpha * know_loss
+        if return_output:
+            return (loss, model_output)
+        else:
+            return loss
+```
+
+### Adding more fields to the batch
+
+Batch is a `namedtuple` containing data sent to an agent. This class is the input type of the train_step and eval_step functions. Agents can override the `batchify` function to return an extended namedtuple with additional fields if they would like. However, we recommend calling the parent function to set up these fields as a base. An example is:
+
+```python
+    def batchify(self, obs_batch):
+        batch = super().batchify(obs_batch)
+        reordered_observations = [obs_batch[i] for i in batch.valid_indices]
+
+        checked_sentences = []
+        for obs in reordered_observations:
```

Comment:
can we use a simpler example. This one has 3 fields and makes it more complicated than it needs to be.